Add coordinator tutorial (#58)
hermeznetwork/docs: docs/users/coordinator.md

# Coordinator
This tutorial focuses on how to launch a Hermez Coordinator node on localhost using the available tools. It includes two parts:
1. [Launching a Coordinator Node](#coordinator-node)
2. [Launching a Proof Server](#proof-server)

## Coordinator Node

> ### Dependencies

Reviewer: I'd recommend to ask the coordinator to compile using Go 1.16 because it has the latest version of crypto adjustments that increase the security.

- golang 1.14+
- [aws cli 2](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html)
- packr utility to bundle the database migrations. Make sure your `$PATH` contains `$GOPATH/bin`, otherwise the packr utility will not be found.
```shell
cd /tmp && go get -u github.com/gobuffalo/packr/v2/packr2 && cd -
```
- docker and docker-compose without sudo permission

### Setup
1. Clone [hermez-node](https://github.com/hermeznetwork/hermez-node.git) repo
```shell
git clone https://github.com/hermeznetwork/hermez-node.git
```
2. Build `hermez-node` executable
```shell
cd hermez-node/db && packr2 && cd ../cli/node
go build .
cd ../../db && packr2 clean && cd -
```
The executable can be found in cli/node/node

> 3. Deploy postgresql database and geth node. For this step we provide a docker-compose file example. Copy contents to file named `docker-compose.sandbox.yaml`

Reviewer: I'd add an alternative to whom does not want to use docker or has AWS dependencies to install GETH using one of these methods: 

- Pre-built binaries for all platforms on our downloads page (https://geth.ethereum.org/downloads/).
- Ubuntu packages in our Launchpad PPA repository (https://launchpad.net/~ethereum/+archive/ubuntu/ethereum).
OSX packages in our Homebrew Tap repository (https://github.com/ethereum/homebrew-ethereum).

And sync this node with Goerli testnet where all Hermez's smart contract are deployed. To do this the only change at GETH's initialization is to change the parameter --dev --goerli

Author: If you don't mind, I wil not add the non-docker installation. If you know what you're doing, you can launch your own ethereum node. I am not getting into this because then i need to go through the contract deployment and i much rather save this for the second phase.
In this second phase, i will describe how to launch a node in testnet (goerli, rinkeby...), using predeployed contracts, and i will talk about this. also i will explain how to launch several coordinators to run the bidding

Reviewer: As per our conversation is worthy to add these lines to help and to be clear to Exchange infra experts they can use directly GETH if they want to instead to use Docker.

More information on Trusted Setup can be found [here](https://github.com/hermeznetwork/phase2ceremony_3).

### Setup
1. Clone [rapidsnark](https://github.com/iden3/rapidsnark.git) repo
```shell
git clone  https://github.com/iden3/rapidsnark.git
```
2. Compile the prover.

In this example we are building the 344 transactions prover.
```shell
cd rapidsnark
npm install
git submodule init
git submodule update
npx task createFieldSources
npx task buildPistche
npx task buildProverServer ../circuit-344-32-256-64.cpp
```
3. Launch prover
```shell
cd ..
./rapidsnark/build/proverServer circuit-344-32-256-64.dat circuit-344-32-256-64_hez3_final.zkey
```
Prover is deployed at port 9080.

4. Check prover status
```shell
curl -i -H "Accept: application/json" -H "Content-Type: application/json" -X GET http://localhost:9080/status
```

### Generate a Prover Input File

- clone repo [circuits](https://github.com/hermeznetwork/circuits.git)
```shell
git clone https://github.com/hermeznetwork/circuits.git
```

- Install dependencies
```shell
cd circuits
npm install
cd tools
```

> In this example we are working with `circuit-344-32-236-64_hez1.zkey`, which correspods to a circuit with 344 transactions, 32 levels, 256 maxL1Tx and 64 maxFeeTx.

Reviewer 2: ```suggestion
In this example, we are working with `circuit-344-32-236-64_hez1.zkey`, which corresponds to a circuit with 344 transactions, 32 levels, 256 maxL1Tx and 64 maxFeeTx.
```

```
This command generates a new input file `rollup-344-32-256-64/input-344-32-256-64.json`

To generate a new input file with random transactions
```shell
node generate-input.js 256 88 344 32 256 64
``` 
This will create a new input file called `inputs-256.json`. 

### Generate a New Proof

You can use curl to post any of the inputs generated in the previous step.
```shell
curl -X POST -d @inputs-256.json http://localhost:9080/input
curl -X POST -d @input-344-32-256-64.json http://localhost:9080/input
```

You check the status of the prover by querting the /status endpoint.
```shell
curl -i -H "Accept: application/json" -H "Content-Type: application/json" -X GET http://localhost:9080/status
```

### Connect Prover to Coordinator Node
Once you have verified the prover is working, you can connect it to the Hermez Coordinator by configuring the `cfg.buidler.toml` configuration file.
You need to substitute sections `ServerProofs` with the updated URL where prover is deployed, and the `Circuit` section where the verifier smart contract to use is specified.

```
[[Coordinator.ServerProofs]]
URL = "http://localhost:9080"
```

```
[Coordinator.Circuit]
MaxTx = 344
NLevels = 32
```

> At this point you can stop the mock server if it is still running, and re-launch the coordinator as we saw in the previous section. The new prover will be running at http://localhost:9080, 

Reviewer 2: ```suggestion
At this point, you can stop the mock server if it is still running, and re-launch the coordinator as we saw in the previous section. The new prover will be running at http://localhost:9080, 
```

> and the two end points are /status and /input

Reviewer 2: ```suggestion
and the two endpoints are /status and /input
```
